pymc3/sampling.py

```diff
@@ -1049,33 +1159,35 @@ def insert(self, k: str, v, idx: int):
             The value of the `idx`th sample from variable `k`
         ids : int
             The index of the sample we are inserting into the trace.
-        '''
-        if hasattr(v, 'shape'):
-            value_shape = tuple(v.shape) # type: Tuple[int, ...]
+        """
+        if hasattr(v, "shape"):
+            value_shape = tuple(v.shape)  # type: Tuple[int, ...]
         else:
             value_shape = ()
 
         # initialize if necessary
         if k not in self.trace_dict:
             array_shape = (self._len,) + value_shape
-            self.trace_dict[k] = np.full(array_shape, np.nan)
+            self.trace_dict[k] = np.zeros(array_shape, dtype=np.array(v).dtype)
```

Comment:
shouldn't we keep `nan` as the default value?

Reply:
Aren't the arrays in `trace_dict` supposed to be completely filled with the posterior predictive samples? Then the filler value would have no relevance at all. In this case, I think the situation would be `np.empty` ideal usecase.

Reply:
Yhis is failing because `nan` are floats so if `trace_dict` is filled with integers they will be upcasted to floats. 

The default value for `empty` is also float so we should specify the dtype as with `zeros`. It seems to me that using `empty` of `zeros` is basically the same, unless using `empty`is preferred because it conveys the idea that the initial values are irrelevant and they will be completetelly replaced. @OriolAbril  Is that the case?

Reply:
Yes to everything, the dtype must be specified and `empty` only allocates the memory that will be occupied by the array so the user can fill the array afterwards without problems (the values of the array may or may not actually be zeros, it depends on what was stored there before). 

Moreover, just for the record because this array allocation is far from being a bottleneck, `empty` also leaves the actual memory untouched, thus `empty` is much faster than `zeros` or `full`.